File: autocard/frontend/src/cad/contracts/document.ts
export type Point = { x: number; y: number }
export type Point3D = { x: number; y: number; z: number }

export type UnitSystem = 'mm' | 'cm' | 'm' | 'inch' | 'ft'

export type DocumentSettings = {
  units: UnitSystem
  precision: number
  angleUnit: 'deg' | 'rad'
  gridSpacing: number
  snapThreshold: number
  defaultLayerId: string
  defaultStyleId: string
  wallDefaultThickness: number
  dimensionStyle: string
}

export type CadDocument = {
  schemaVersion: number
  documentId: string
  name: string
  createdAt: string
  updatedAt: string
  units: UnitSystem
  settings: DocumentSettings
  roots: string[]
  nodes: Record<string, import('./nodes').CadNode>
  layers: Record<string, import('./layers').LayerDef>
  styles: Record<string, import('./layers').StyleDef>
  blocks: Record<string, import('./nodes/composition').BlockDefinitionNode>
  views: Record<string, import('./sheets').ViewDef>
  sheets: Record<string, import('./sheets').SheetDef>
  constraints: Record<string, import('./constraints').ConstraintDef>
  derived?: DerivedDocumentState
  metadata?: Record<string, unknown>
}

export type WallPolygonCache = {
  nodeId: string
  outline: Point[]
  centerLine: Point[]
  thickness: number
  joinsWith: string[]
}

export type WallJoinCache = {
  nodeId: string
  joinTypeByNeighbor: Record<string, 'miter' | 'butt' | 'tee' | 'none'>
  intersectionPoints: Point[]
}

export type RoomGraphCache = {
  nodeId: string
  boundary: Point[]
  area: number
  perimeter: number
  wallIds: string[]
  openingIds: string[]
  roomType?: string
}

export type RoomLabelCache = {
  nodeId: string
  label: string
  position: Point
  angle?: number
}

export type OpeningPlacementCache = {
  nodeId: string
  hostWallId: string
  position: Point
  wallTangent: Point
  wallNormal: Point
  cutWidth: number
}

export type BoundsCache = {
  nodeId: string
  minX: number
  minY: number
  maxX: number
  maxY: number
}

export type LayerVisibilityCache = {
  layerId: string
  visible: boolean
  effectiveNodeIds: string[]
}

export type SnapPoint = {
  position: Point
  type: 'endpoint' | 'midpoint' | 'center' | 'intersection' | 'perpendicular' | 'tangent' | 'grid' | 'nearest'
  nodeId?: string
  priority: number
}

export type SnapIndexCache = {
  revision: number
  endpoints: SnapPoint[]
  midpoints: SnapPoint[]
  centers: SnapPoint[]
  intersections: SnapPoint[]
  gridPoints?: SnapPoint[]
}

export type SpatialBucket = {
  bucketKey: string
  nodeIds: string[]
}

export type SpatialIndexCache = {
  revision: number
  buckets: Record<string, string[]>
  boundsByNode: Record<string, BoundsCache>
}

export type RenderIndexCache = {
  revision: number
  drawOrder: string[]
  selectableOrder: string[]
  hoverPriority: string[]
}

export type DerivedDiagnostic = {
  code: string
  severity: 'info' | 'warning' | 'error'
  nodeIds: string[]
  message: string
}

export type DerivedDocumentState = {
  revision: number
  fromDocumentVersion: number
  wallPolygons: Record<string, WallPolygonCache>
  wallJoins: Record<string, WallJoinCache>
  roomGraphs: Record<string, RoomGraphCache>
  roomLabels: Record<string, RoomLabelCache>
  openingPlacements: Record<string, OpeningPlacementCache>
  nodeBounds: Record<string, BoundsCache>
  layerVisibility: Record<string, LayerVisibilityCache>
  snapIndex?: SnapIndexCache
  spatialIndex?: SpatialIndexCache
  renderIndex?: RenderIndexCache
  diagnostics: DerivedDiagnostic[]
}

File: autocard/backend/cad/schema/document.go
// Package schema defines the canonical CAD document model.
// These structs mirror the TypeScript types in cad/contracts/document.ts.
// When TS types change, update this file in the same commit.
package schema

import "encoding/json"

type Point struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type UnitSystem string

const (
	UnitMM   UnitSystem = "mm"
	UnitCM   UnitSystem = "cm"
	UnitM    UnitSystem = "m"
	UnitInch UnitSystem = "inch"
	UnitFt   UnitSystem = "ft"
)

type DocumentSettings struct {
	Units                 UnitSystem `json:"units"`
	Precision             int        `json:"precision"`
	AngleUnit             string     `json:"angleUnit"`
	GridSpacing           float64    `json:"gridSpacing"`
	SnapThreshold         float64    `json:"snapThreshold"`
	DefaultLayerID        string     `json:"defaultLayerId"`
	DefaultStyleID        string     `json:"defaultStyleId"`
	WallDefaultThickness  float64    `json:"wallDefaultThickness"`
	DimensionStyle        string     `json:"dimensionStyle"`
}

// CadDocument mirrors cad/contracts/document.ts CadDocument.
type CadDocument struct {
	SchemaVersion int                        `json:"schemaVersion"`
	DocumentID    string                     `json:"documentId"`
	Name          string                     `json:"name"`
	CreatedAt     string                     `json:"createdAt"`
	UpdatedAt     string                     `json:"updatedAt"`
	Units         UnitSystem                 `json:"units"`
	Settings      DocumentSettings           `json:"settings"`
	Roots         []string                   `json:"roots"`
	Nodes         map[string]json.RawMessage `json:"nodes"`
	Layers        map[string]LayerDef        `json:"layers"`
	Styles        map[string]StyleDef        `json:"styles"`
	Blocks        map[string]json.RawMessage `json:"blocks"`
	Views         map[string]ViewDef         `json:"views"`
	Sheets        map[string]SheetDef        `json:"sheets"`
	Constraints   map[string]ConstraintDef   `json:"constraints"`
	Derived       *DerivedDocumentState      `json:"derived,omitempty"`
	Metadata      map[string]interface{}     `json:"metadata,omitempty"`
}

// DerivedDocumentState mirrors cad/contracts/document.ts DerivedDocumentState.
type DerivedDocumentState struct {
	Revision            int                             `json:"revision"`
	FromDocumentVersion int                             `json:"fromDocumentVersion"`
	WallPolygons        map[string]WallPolygonCache     `json:"wallPolygons"`
	WallJoins           map[string]WallJoinCache        `json:"wallJoins"`
	RoomGraphs          map[string]RoomGraphCache       `json:"roomGraphs"`
	RoomLabels          map[string]RoomLabelCache       `json:"roomLabels"`
	OpeningPlacements   map[string]OpeningPlacementCache `json:"openingPlacements"`
	NodeBounds          map[string]BoundsCache          `json:"nodeBounds"`
	LayerVisibility     map[string]LayerVisibilityCache `json:"layerVisibility"`
	SnapIndex           *SnapIndexCache                 `json:"snapIndex,omitempty"`
	SpatialIndex        *SpatialIndexCache              `json:"spatialIndex,omitempty"`
	RenderIndex         *RenderIndexCache               `json:"renderIndex,omitempty"`
	Diagnostics         []DerivedDiagnostic             `json:"diagnostics"`
}

type WallPolygonCache struct {
	NodeID     string    `json:"nodeId"`
	Outline    []Point   `json:"outline"`
	CenterLine []Point   `json:"centerLine"`
	Thickness  float64   `json:"thickness"`
	JoinsWith  []string  `json:"joinsWith"`
}

type WallJoinCache struct {
	NodeID              string             `json:"nodeId"`
	JoinTypeByNeighbor  map[string]string  `json:"joinTypeByNeighbor"`
	IntersectionPoints  []Point            `json:"intersectionPoints"`
}

type RoomGraphCache struct {
	NodeID     string   `json:"nodeId"`
	Boundary   []Point  `json:"boundary"`
	Area       float64  `json:"area"`
	Perimeter  float64  `json:"perimeter"`
	WallIDs    []string `json:"wallIds"`
	OpeningIDs []string `json:"openingIds"`
	RoomType   string   `json:"roomType,omitempty"`
}

type RoomLabelCache struct {
	NodeID   string  `json:"nodeId"`
	Label    string  `json:"label"`
	Position Point   `json:"position"`
	Angle    float64 `json:"angle,omitempty"`
}

type OpeningPlacementCache struct {
	NodeID      string  `json:"nodeId"`
	HostWallID  string  `json:"hostWallId"`
	Position    Point   `json:"position"`
	WallTangent Point   `json:"wallTangent"`
	WallNormal  Point   `json:"wallNormal"`
	CutWidth    float64 `json:"cutWidth"`
}

type BoundsCache struct {
	NodeID string  `json:"nodeId"`
	MinX   float64 `json:"minX"`
	MinY   float64 `json:"minY"`
	MaxX   float64 `json:"maxX"`
	MaxY   float64 `json:"maxY"`
}

type LayerVisibilityCache struct {
	LayerID         string   `json:"layerId"`
	Visible         bool     `json:"visible"`
	EffectiveNodeIDs []string `json:"effectiveNodeIds"`
}

type SnapPoint struct {
	Position Point   `json:"position"`
	Type     string  `json:"type"`
	NodeID   string  `json:"nodeId,omitempty"`
	Priority int     `json:"priority"`
}

type SnapIndexCache struct {
	Revision      int         `json:"revision"`
	Endpoints     []SnapPoint `json:"endpoints"`
	Midpoints     []SnapPoint `json:"midpoints"`
	Centers       []SnapPoint `json:"centers"`
	Intersections []SnapPoint `json:"intersections"`
	GridPoints    []SnapPoint `json:"gridPoints,omitempty"`
}

type SpatialIndexCache struct {
	Revision     int                    `json:"revision"`
	Buckets      map[string][]string    `json:"buckets"`
	BoundsByNode map[string]BoundsCache `json:"boundsByNode"`
}

type RenderIndexCache struct {
	Revision        int      `json:"revision"`
	DrawOrder       []string `json:"drawOrder"`
	SelectableOrder []string `json:"selectableOrder"`
	HoverPriority   []string `json:"hoverPriority"`
}

type DerivedDiagnostic struct {
	Code     string   `json:"code"`
	Severity string   `json:"severity"`
	NodeIDs  []string `json:"nodeIds"`
	Message  string   `json:"message"`
}
